src/user-guide/writing-workflows/optional-outputs.rst

```diff
@@ -0,0 +1,266 @@
+
+.. _User Guide Optional Outputs:
+
+Expected & Optional Task Outputs
+================================
+
+Cylc 8 follows the graph wherever it leads, as events dictate at runtime. This
+is powerful, but it is not necessarily helpful if tasks behave in ways not
+anticipated by the graph.
+
+For instance, the following graph says that task ``bar`` should trigger if
+``foo`` succeeds:
+
+.. code-block:: cylc
+
+   [graph]
+       R1 = "foo => bar"
+
+If ``foo`` actually fails at runtime, however, and the graph does not recognize
+that as a possible outcome, the scheduler should probably conclude that the
+workflow did not run to completion as expected, rather than shutting down
+just because the graph says there is nothing else to do.
+
+It may not be sufficient just to highlight the failure of ``foo`` as a problem,
+because sometimes task failure is expected, and other task outputs can also
+lead to dead-end side branches of the graph.
+
+So, to allow proper diagnosis of workflow completion, the scheduler *expects*
+all task outputs to be completed at runtime, unless they  are explicitly marked
+as *optional*.
+
+
+Incomplete Tasks
+----------------
+
+Tasks that finish without completing expected outputs [1]_ are retained as
+:term:`incomplete tasks <incomplete task>` pending user intervention (e.g. to
+be retriggered after fixing the problem, to allow the workflow to continue).
+
+.. note::
+   Incomplete tasks count toward the :term:`runahead limit`.
+
+
+Scheduler Stall
+---------------
+
+If there is nothing else to do, the scheduler will conclude that the workflow
+has run to completion, and shut down, if there are no incomplete tasks present.
+
+But if there are any incomplete tasks present, the scheduler will log a
+:term:`stall` and stay alive for 1 hour (by default) to allow user intervention.
+
+Restarting a stalled workflow will trigger a new stall timer.
+
+.. note::
+   
+   Partially satisfied prerequisites can also cause a stall. If ``a & b => c``,
+   but only ``a`` runs (perhaps something went wrong upstream of ``b``) the
+   scheduler will take partial completion of ``c``'s prerequisites as a sign
+   that the workflow did not run to completion as expected.
+
+Examples
+--------
+
+Outputs are expected by default (and the tasks will be retained as incomplete
+if they are not generated):
+
+.. code-block:: cylc
+
+   [graph]
+       R1 = """
+          foo  # foo:succeed expected
+          bar:x  # bar:x expected
+       """
+
+Outputs can be flagged as optional with a question mark:
+
+.. code-block:: cylc
+
+   [graph]
+       R1 = """
+          foo?  # foo:succeed optional
+          bar:x?  # bar:x optional
+       """
+
+.. warning::
+
+   To avoid confusion, optional outputs must be marked as such wherever they
```

Comment:
```suggestion
   To avoid confusion, optional outputs must be marked as such every single time they
```

Right?

It might be useful to illustrate this with an example

Reply:
(To my mind, that's what "wherever they occur" means, but fair enough).